### PR TITLE
[DNMY] Improve input validation in order to figure out avg/min mismatch.

### PR DESCRIPTION
**Pull request by jana-rapava**
Signed-off-by: Jana Rapava <janarapava@eaton.com>

### CODE REVIEW

File: src/cmstats.c [hunk 1]
```diff
@@ -49,6 +50,26 @@ s_duplicator (const void *self)
     return (void*) fty_proto_dup ((fty_proto_t*) self);
 }
 
+static double
+safe_atof (const char *str)
+{
+    char *endptr = (char *) zmalloc (sizeof (char));
+    errno = 0;
+    double ret = strtod (str, &endptr);
```
Comment (jimklimov) [inline src/cmstats.c]: This looks fishy: according to common logic of such functions (and UPDATE: docs at https://www.tutorialspoint.com/c_standard_library/c_function_strtod.htm too) the `endptr` should just be an (uninitialized) `char *` where the `strtod()` returns a NULL or pointer to the non-numeric character from the tail of `str`. I guess if all `str` was processed into a number, the `endptr` will point to the `\0` of `str`.

In short, no `malloc()` and no `free()` are needed.

File: src/cmstats.c [hunk 2]
```diff
@@ -49,6 +50,26 @@ s_duplicator (const void *self)
...
+    // str does not contain number
+    if (ret == 0 && endptr == str) {
```
Comment (jimklimov) [inline src/cmstats.c]: Comparison `ret == 0.0` seems cleaner, to stress (to compiler too) the floatiness of this zero :)

File: src/cmstats.c [hunk 3]
```diff
@@ -49,6 +50,26 @@ s_duplicator (const void *self)
...
+        //ret = "NAN";
+    }
+    // overflow or underflow
+    if (ret == 0 && errno == ERANGE) {
+        log_error ("strtod experienced over/underflow when converting %s", str);
+        //ret = "NAN";
+    }
+
+    free (endptr);
+    return ret;
```
Comment (jimklimov) [inline src/cmstats.c]: Should we return the zero number, like before? Is the added benefit here at the moment the logging in case of error, so we can at least have some debug log when issues happen? Maybe also log the string at `endptr` if it is not NULL and `endptr[0] != '\0'` to see if some wrong text got into messages?

**Comment by jana-rapava** [comment]
@jimklimov : I made suggested changes - now only Jenkins memcheck-with-DRAFTS is failing and I can't reproduce the issue locally. Any ideas?

**Comment by jimklimov** [comment]
@jana-rapava : The actual error message in that test was an assertion fail... hard to read, lots of expansions involved :)

The rest of the log is about how valgrind saw unfreed memory (because assertion aborted the program uncleanly).

I failed to split/indent that huge line in finite time... but it was in https://github.com/42ity/fty-metric-compute/blob/master/src/fty_mc_server.c#L504 so I guess you should check if the codepath tested there is anyhow related to what you were changing here? (And/or if you found a way to get some race, lag or otherwise trigger our bug?)

lt-fty_metric_compute_selftest: src/fty_mc_server.c:504: fty_mc_server_test: Assertion `(!__extension__ ({ size_t __s1_len, __s2_len; (__builtin_constant_p ((fty_proto_value (bmsg))) && __builtin_constant_p (("50.00")) && (__s1_len = __builtin_strlen ((fty_proto_value (bmsg))), __s2_len = __builtin_strlen (("50.00")), (!((size_t)(const void *)(((fty_proto_value (bmsg))) + 1) - (size_t)(const void *)((fty_proto_value (bmsg))) == 1) || __s1_len >= 4) && (!((size_t)(const void *)((("50.00")) + 1) - (size_t)(const void *)(("50.00")) == 1) || __s2_len >= 4)) ? __builtin_strcmp ((fty_proto_value (bmsg)), ("50.00")) : (__builtin_constant_p ((fty_proto_value (bmsg))) && ((size_t)(const void *)(((fty_proto_value (bmsg))) + 1) - (size_t)(const void *)((fty_proto_value (bmsg))) == 1) && (__s1_len = __builtin_strlen ((fty_proto_value (bmsg))), __s1_len < 4) ? (__builtin_constant_p (("50.00")) && ((size_t)(const void *)((("50.00")) + 1) - (size_t)(const void *)(("50.00")) == 1) ? __builtin_strcmp ((fty_proto_value (bmsg)), ("50.00")) : (__extension__ ({ const unsigned char *__s2 = (const unsigned char *) (const char *) (("50.00")); int __result = (((const unsigned char *) (const char *) ((fty_proto_value (bmsg))))[0] - __s2[0]); if (__s1_len > 0 && __result == 0) { __result = (((const unsigned char *) (const char *) ((fty_proto_value (bmsg))))[1] - __s2[1]); if (__s1_len > 1 && __result == 0) { __result = (((const unsigned char *) (const char *) ((fty_proto_value (bmsg))))[2] - __s2[2]); if (__s1_len > 2 && __result == 0) __result = (((const unsigned char *) (const char *) ((fty_proto_value (bmsg))))[3] - __s2[3]); } } __result; }))) : (__builtin_constant_p (("50.00")) && ((size_t)(const void *)((("50.00")) + 1) - (size_t)(const void *)(("50.00")) == 1) && (__s2_len = __builtin_strlen (("50.00")), __s2_len < 4) ? (__builtin_constant_p ((fty_proto_value (bmsg))) && ((size_t)(const void *)(((fty_proto_value (bmsg))) + 1) - (size_t)(const void *)((fty_proto_value (bmsg))) == 1) ? __builtin_strcmp ((fty_proto_value (bmsg)), ("50.00")) : (- (__extension__ ({ const unsigned char *__s2 = (const unsigned char *) (const char *) ((fty_proto_value (bmsg))); int __result = (((const unsigned char *) (const char *) (("50.00")))[0] - __s2[0]); if (__s2_len > 0 && __result == 0) { __result = (((const unsigned char *) (const char *) (("50.00")))[1] - __s2[1]); if (__s2_len > 1 && __result == 0) { __result = (((const unsigned char *) (const char *) (("50.00")))[2] - __s2[2]); if (__s2_len > 2 && __result == 0) __result = (((const unsigned char *) (const char *) (("50.00")))[3] - __s2[3]); } } __result; })))) : __builtin_strcmp ((fty_proto_value (bmsg)), ("50.00"))))); }))' failed.

File: src/cmstats.c [hunk 4]
```diff
@@ -49,6 +50,25 @@ s_duplicator (const void *self)
...
 }
 
+static double
+safe_atof (const char *str)
+{
+    char *endptr;
+    errno = 0;
+    double ret = strtod (str, &endptr);
+    // we could not process whole str
+    if (ret == 0.0 && endptr != NULL && endptr[0] != '\0') {
```
Comment (jimklimov) [inline src/cmstats.c]: Technically, an input string like `0.000000abc` would also cause this - with the number part really being zero. But in our use-case it should not happen anyway. :)

File: src/cmstats.c [hunk 5]
```diff
@@ -49,6 +50,25 @@ s_duplicator (const void *self)
...
+        log_error ("strtod received %s with dangling non-number argument '%s'", str, endptr);
+        //ret = "NAN";
+    }
+    // overflow or underflow
+    if (errno == ERANGE) {
```
Comment (jimklimov) [inline src/cmstats.c]: Should we complain about actually any non-zero `errno`? Should `strtod()` reset it to okay state if all went well?

**Comment by jimklimov** [comment]
@jana-rapava : note that https://github.com/42ity/fty-metric-compute/pull/30 was merged causing some conflicts; in particular it bumped test timings from 1s to 10s (so maybe working around that failure - or hiding why it sometimes goes wrong... now the bug would happen with 10% of the already low probability...)

**Comment by jana-rapava** [comment]
@jimklimov : Based on Jenkins builds, now if it fails, it fails in completely different part of the test :-)
